### PR TITLE
Added support for BlackWidow V3 Pro (Wired)

### PR DESCRIPTION
**Pull request by geodox**
Fixes part of #1279

All credit goes to @faridgh1991 for his PR #1514

As the PR went stale, I decided it would be better to push the support for wired mode while wireless and bluetooth can be worked on in the future (something I'm going to try to do next).

To build and test
https://github.com/openrazer/openrazer/wiki/Building-a-package

### CODE REVIEW

**Comment by GeoDoX** [comment]
Failing the check-formatting test, tried auto formatting but getting `xargs: astyle: No such file or directory`.

**Comment by IvanVojtko** [comment]
I guess that wireless mode will be same as it is on my mini version. You can check my PR https://github.com/openrazer/openrazer/pull/1617. Basically you just need to use `9F` instead of `1F` in every packet.

**Comment by GeoDoX** [comment]
@IvanVojtko For the pro in Wired mode, some packets are `3F` and some are `1F`. For the mini it seems they're all `1F` in wired mode. Don't wanna chance it till I can get some pcaps as I'm fairly new to deving for this. I'm open to a conversation about it though.

**Comment by EduardTheFirst** [comment]
> Fixes #1619
> 
> All credit goes to @faridgh1991 for his PR #1514
> 
> As the PR went stale, I decided it would be better to push the support for wired mode while wireless and bluetooth can be worked on in the future (something I'm going to try to do next).
> 
> To build and test https://github.com/openrazer/openrazer/wiki/Building-a-package

hey do you have discord ? 
or telegram ?

**Comment by geodox** [comment]
@EduardTheFirst I am on the openrazer telegram. I also have discord geodox#1051

**Comment by geodox** [comment]
Don’t forget to comment back with your results of testing! It helps get this merged into official.

Positive and negative!

**Comment by EduardTheFirst** [comment]
> @EduardTheFirst I am on the openrazer telegram. I also have discord geodox#1051

I have added you on discord but i haven't found you on telegram

File: install_files/udev/99-razer.rules [hunk 1]
```diff
@@ -10,7 +10,7 @@ ATTRS{idProduct}=="0013|0016|0020|0024|0025|002e|002f|0032|0034|0036|0037|0038|0
     ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
 
 # Keyboards
-ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0258|0259|025d|025e|0266|026a|026d|0270|0271|0276|0a24", \
+ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0257|0258|0259|025a|025d|025e|0266|026a|026d|0276|0a24", \
```
Comment (z3ntu) [inline install_files/udev/99-razer.rules]: Please don't remove `0270|0271` from the list

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -345,32 +346,6 @@ static void razer_set_device_mode(struct usb_device *usb_dev, unsigned char mode
...
-        break;
-    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI_WIRELESS:
-        report.transaction_id.id = 0x9f;
-        break;
-    }
-
-    response_report = razer_send_payload(usb_dev, &report);
-
-    return sprintf(buf, "%d\n", response_report.arguments[1]);
-}
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Is there any reason you're moving this function down? If not please keep it here (and add the others near this as well)

File: README.md [hunk 3]
```diff
@@ -71,6 +71,7 @@ The devices below are fully feature supported by OpenRazer, which means all avai
 | Razer Huntsman Mini                           |  1532:0257  |
 | Razer BlackWidow V3 Mini Hyperspeed           |  1532:0258  |
 | Razer Blade Stealth (Late 2020)               |  1532:0259  |
+| Razer BlackWidow V3 (Wired)                   |  1532:025A  |
```
Comment (z3ntu) [inline README.md]: Isn't this device the BlackWidow V3 Pro?

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 4]
```diff
@@ -359,6 +359,37 @@ class RazerBlackWidowTournamentEdition2014(_MacroKeyboard):
     DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/products/17564/razer-blackwidow-te-stealth-hero-01.png"
 
 
+class RazerBlackWidowV3ProWired(_RippleKeyboard):
+    """
+    Class for the Razer BlackWidow V3 Pro Wired
+    """
+    DEVICE_NAME = 'Razer BlackWidow V3 Pro Wired'
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: Pleases don't declare custom variables, and please use brackets around the "Wired"

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -726,6 +701,10 @@ static ssize_t razer_attr_read_device_type(struct device *dev, struct device_att
         device_type = "Razer Huntsman Mini\n";
         break;
 
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRED:
+        device_type = "Razer BlackWidow V3 Pro(Wired)\n";
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: add a space before the bracket

**Review by z3ntu — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (phl0) [reply README.md]: VID/PID match my Black Window V3 Pro. So I think it should be renamed accordingly.

Comment (Noah-Leick) [reply README.md]: @phl0 Pending review from z3ntu, the issue with naming is fixed with commit 484f51a74c822599182e61ce7b0a6aad1e857fa9.

**Comment by phl0** [comment]
I did test the current code of this PR/branch and it works. I can change colors on my Black Widow V3 pro. So besides the code beauty this would be nice to be merged upstream imho.

DSC_2002

Screenshot from 2021-10-20 12-22-05

**Comment by z3ntu** [comment]
I've fixed up some things and rebased, can you please try the latest commit here again to see if I haven't broken anything?

**Comment by wyattearp** [comment]
FYI - just another rando on the internet that's built this PR & tested successfully:

image

Other notes:
- Ubuntu 20.04.3 LTS
- ./scripts/build_debs.sh used installation

**Comment by wyattearp** [comment]
FYI - the only conflict I'm seeing on the merge is the fact that both the wired / wireless branches are stomping on the same spot in the code (because of course they are :) ).

```diff
diff --cc driver/razerkbd_driver.c
index 8903e56,c569ca0..0000000
--- a/driver/razerkbd_driver.c
+++ b/driver/razerkbd_driver.c
@@@ -853,8 -783,8 +865,13 @@@ static ssize_t razer_attr_read_device_t
          device_type = "Razer Huntsman Mini\n";
          break;
  
++<<<<<<< HEAD
 +    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRED:
 +        device_type = "Razer BlackWidow V3 Pro (Wired)\n";
++=======
+     case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3:
+         device_type = "Razer BlackWidow V3\n";
++>>>>>>> master
          break;
  
```

Conflict 2:
```
diff --cc install_files/udev/99-razer.rules
index 0b3150a,9397c75..0000000
--- a/install_files/udev/99-razer.rules
+++ b/install_files/udev/99-razer.rules
@@@ -10,7 -10,7 +10,11 @@@ ATTRS{idProduct}=="0013|0016|0020|0024|
      ENV{ID_RAZER_CHROMA}="1", ENV{RAZER_DRIVER}="razermouse"
  
  # Keyboards
++<<<<<<< HEAD
 +ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|0252|0253|0255|0256|0257|0258|0259|025a|025d|025e|0266|026a|026d|0270|0271|0276|0279|0a24", \
++=======
+ ATTRS{idProduct}=="010d|010e|010f|0111|0113|011a|011b|011c|0201|0202|0203|0204|0205|0207|0208|0209|020f|0210|0211|0214|0216|0217|021a|021e|021f|0220|0221|0224|0225|0226|0227|0228|022a|022b|022c|022d|022f|0232|0233|0234|0235|0237|0239|023a|023b|023f|0240|0241|0243|0245|0246|024a|024c|024d|024e|0252|0253|0255|0256|0257|0258|0259|025d|025e|0266|026a|026d|0270|0271|0276|0279|0a24", \
++>>>>>>> master
```

Just don't forget the second `break;` in the driver and the addition of the second `ATTRS` line on the merge (because they're the same they get overlooked by both a rebase and a merge strategy).

**Comment by amurp** [comment]
I see there are conflicts, but I'm dropping in to say this works for me once built:

image

**Comment by eeyrjmr** [comment]
working here as well 


Gentoo, wayland, clang

I had to refactor the gentoo ebuild (not just to use this PR but generally). That is a gentoo issue not openrazer/PR 

good job

**Comment by wyattearp** [comment]
Bump - it's currently a very clean rebase/merge if you can hit

**Comment by charlydelta** [comment]
Is there anything hindering the merge? What's the status? I'd really love to see this merged soon, please don't let this bit-rot again.

**Review by z3ntu — APPROVED** [review]
LGTM, thanks for fixing the remaining issues and to everyone for testing!

**Pull request merged**